### PR TITLE
Updated Crypto and CipherParams

### PR DESCRIPTION
**Pull request by bladeSk**
Implements the enhancement https://github.com/ably/ably-php/issues/29

### CODE REVIEW

File: src/Channel.php [hunk 1]
```diff
@@ -27,7 +27,7 @@ class Channel {
      * Constructor
      * @param AblyRest $ably Ably API instance
      * @param string $name Channel's name
-     * @param array $options Channel options ['encrypted', 'cipherParams']
+     * @param ChannelOpions|array|null $options Channel options (for encrypted channels)
```
Comment (mattheworiordan) [inline src/Channel.php]: Typo, should be ChannelOptions

File: src/Utils/Crypto.php [hunk 2]
```diff
@@ -41,11 +42,59 @@ public static function decrypt( $payload, $cipherParams ) {
...
+
+        if ( isset( $params['base64Key'] ) && $params['base64Key'] ) {
+            $params['key'] = strtr( $params['key'], '_-', '/+' );
+            $params['key'] = base64_decode( $params['key'] );
+        }
+
+        $cipherParams->key = $params['key'];
+        $cipherParams->keyLength = isset( $params['keyLength'] ) ? $params['keyLength'] : strlen( $cipherParams->key ) * 8;
+        
+        if ( !in_array( $cipherParams->keyLength, array( 128, 256 ) ) ) {
```
Comment (mattheworiordan) [inline src/Utils/Crypto.php]: This only applies if the algorithm is `aes`, if not, we don't enforce this even though we're not sure what other algorithms may or may not work

File: src/Utils/Crypto.php [hunk 3]
```diff
@@ -41,11 +42,59 @@ public static function decrypt( $payload, $cipherParams ) {
...
+        if ( $cipherParams->keyLength / 8 != strlen( $cipherParams->key ) ) {
+            throw new AblyException ( 'keyLength does not match the actual key length.', 40003, 400 );
+        }
+
+        $cipherParams->algorithm = isset( $params['algorithm'] ) ? $params['algorithm'] : 'aes';
+        $cipherParams->mode = isset( $params['mode'] ) ? $params['mode'] : 'cbc';
+
+        if ( !in_array( $cipherParams->getAlgorithmString(), array( 'aes-128-cbc', 'aes-256-cbc' ) ) ) {
+            throw new AblyException ( 'Unsupported cipher configuration "' . $cipherParams->getAlgorithmString()
+                . '". The supported configurations are aes-128-cbc and aes-256-cbc', 40003, 400 );
```
Comment (mattheworiordan) [inline src/Utils/Crypto.php]: Are we sure other algorithms aren't supported if they specified it?  We've taken the view in other libs that we don't block it allowing the user to configure the cipher at their peril

**Comment by mattheworiordan** [comment]
Few comments but otherwise LGTM

File: src/Utils/Crypto.php (hunk 3 above)
Comment (bladeSk) [inline src/Utils/Crypto.php]: I thought we enforce only these 2 particular combinations. Didn't know about the change. Added support for any algorithm supported by openssl and a test to check if something else besides AES works.

**Pull request merged**